zlib: add ArrayBuffer support

JemBijoux opened the pull request:
Refs: https://github.com/nodejs/node/issues/1826
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
zlib, src

TimothyGu commented on doc/api/zlib.md:
```diff
@@ -308,7 +311,7 @@ ignored by the decompression classes.
 * `memLevel` {integer} (compression only)
 * `strategy` {integer} (compression only)
-* `dictionary` {Buffer|TypedArray|DataView} (deflate/inflate only, empty dictionary by
+* `dictionary` {Buffer|TypedArray|DataView|ArrayBuffer} (deflate/inflate only, empty dictionary by
```
Looks like this line might exceed 80 characters... either you can re-wrap, or we can do it while landing this.

TimothyGu commented on src/util.h:
```diff
@@ -418,13 +418,22 @@ class BufferValue : public MaybeStackBuffer<char> {
+    v8::ArrayBuffer::Contents name##_c = name->GetContents();                 \
+    name##_length = name##_c.ByteLength();                                    \
+    name##_data =                                                             \
+        static_cast<char*>(name##_c.Data());                                  \
```
These two lines can be joined into one. Again, it would be great if you could fix that, but if not we can fix that while landing.

cjihrig commented on lib/zlib.js:
```diff
@@ -238,7 +241,7 @@ function Zlib(opts, mode) {
 
     dictionary = opts.dictionary;
-    if (dictionary !== undefined && !isArrayBufferView(dictionary)) {
+    if (dictionary !== undefined && !isArrayBufferView(dictionary) && !isAnyArrayBuffer(dictionary)) {
```
This line needs to be wrapped at 80 characters as well.

addaleax commented:
CI: https://ci.nodejs.org/job/node-test-commit/12848/

JemBijoux commented:
I addressed those couple of line length issues. Is there a code style guide somewhere that I should reference in the future? (I just ran my changes through eslint to get an idea of what to change.)

@addaleax did you link to some CI issues there? Looks like some failures. I had a glance but am not certain about the output there. 

Also: when you suggested splitting were you talking the C++ code for the ArrayBuffers? Big thanks to @TimothyGu for holding my hand during my updating those; however I thought the work was pretty interrelated? If you want that part split, can it happen "during landing" as you suggested.

addaleax commented:
@JemBijoux I started CI jobs for this, yes. The failures look like this:

```js
    assert.js:45
      throw new errors.AssertionError({
      ^
    
    AssertionError [ERR_ASSERTION]: 'The "buffer" argument must be one of type string, Buffer, TypedArray, DataView, or ArrayBuffer' === 'The "buffer" argument must be one of type string, Buffer, TypedArray, or DataView'
        at Object.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux/nodes/ubuntu1610-x64/test/common/index.js:723:16)
        at Object.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux/nodes/ubuntu1610-x64/test/common/index.js:517:15)
        at expectedException (assert.js:651:19)
        at innerThrows (assert.js:685:21)
        at Function.throws (assert.js:702:3)
        at Object.expectsError (/home/iojs/build/workspace/node-test-commit-linux/nodes/ubuntu1610-x64/test/common/index.js:745:12)
        at __dirname.forEach (/home/iojs/build/workspace/node-test-commit-linux/nodes/ubuntu1610-x64/test/parallel/test-zlib-not-string-or-buffer.js:10:10)
        at Array.forEach (<anonymous>)
        at Object.<anonymous> (/home/iojs/build/workspace/node-test-commit-linux/nodes/ubuntu1610-x64/test/parallel/test-zlib-not-string-or-buffer.js:9:65)
        at Module._compile (module.js:600:30)
```

Can you reproduce those by running `make test`?

And, what I’m basically thinking is that the changes to `src/node_buffer.cc` could be in their own commit, because they are worth calling out explicitly in the changelog: They change the API for all native addon authors, and that’s not really obvious if the commit is referring only to `zlib`. That’s basically all. :)

JemBijoux commented:
Hey! So I was able to confirm the issue with the test and make that fix no problem. All tests passing again.

I also moved that node_buffer.cc file into it's own commit now (https://github.com/nodejs/node/pull/16111). However, when I did that, one of the changes to a test I had here started to fail also in `test/parallel/test-zlib-convenience-methods.js` so I've removed that change for now.

lpinca commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/10711/

BridgeAR commented:
I think this is ready to land. Ping @TimothyGu

addaleax commented:
@BridgeAR See also the discussion in https://github.com/nodejs/node/pull/16111 … I am not quite at the point where I think extending the Buffer code for this is necessarily a great idea …

I think for now it *might* be nicer to let this be a zlib-only thing and create an Uint8Array for passed ArrayBuffers, then work with that, if we want support?

BridgeAR commented:
I see. In that case I just mark this accordingly for now.

Trott commented:
I'm a little unsure what the situation is here. What would have to happen for this to get unblocked?

Trott commented:
> I'm a little unsure what the situation is here. What would have to happen for this to get unblocked?

Ping @BridgeAR @addaleax

gireeshpunathil commented:
ping @BridgeAR and @addaleax again to have another look ( one of 3 remaining `code-and-learn` PRs!)

addaleax commented:
This is blocked on #16111, that’s all.

I think for zlib this might make more sense than for the general case of all Buffer-taking APIs (although “compress a chunk of memory” is still going to be exceedingly rare compared to “compress a chunk of data”, conceptually).

It might be a nicer approach to create a `Uint8Array` over it when an `ArrayBuffer` is passed to the API? @TimothyGu ?

TimothyGu commented:
@addaleax That works too!

TimothyGu commented:
I've rebased this patch to be minimally invasive with regards to public APIs. @addaleax, @jasnell, @joyeecheung, @lpinca, @BridgeAR, @cjihrig, PTAL.

CI: https://ci.nodejs.org/job/node-test-pull-request/11917/

addaleax commented:
Landed in 284871862ed0fc08ed2946be9ee005c16877d05d 🎉

MylesBorins commented:
Would we want to backport to v8.x? If so this will need a manual backport
